core/src/test/java/apoc/nodes/NodesTest.java

```diff
@@ -57,6 +61,65 @@ public void isDense() throws Exception {
+                "WITH m1, m2, m3 \n" +
+                "CREATE (m1)-[:DEPENDS_ON {id: 3}]->(m2), (m2)-[:ANOTHER {id: 4}]->(m3), (m2)-[:DEPENDS_ON {id: 5}]->(m3)\n" +
+                "CREATE (m1)-[:DEPENDS_ON {id: 6}]->(:Module {bar: 'seven'})-[:DEPENDS_ON {id: 7}]->(:Module {bar: 'eight'})-[:DEPENDS_ON {id: 8}]->(m1)");
+
+        // node with 1 'MY_REL' cycle
+        db.executeTransactionally("CREATE (m1:Start {bar: 'beta'}) with m1\n" +
+                "CREATE (m1)-[:MY_REL {id: 9}]->(m2:Module {bar: 'three'})-[:MY_REL  {id: 10}]->(m3:Module {bar: 'four'})-[:MY_REL {id: 11}]->(m1)");
+
+        // node without cycle
+        db.executeTransactionally("CREATE (m1:Start {bar: 'gamma'}) with m1\n" +
+                "CREATE (m1)-[:DEPENDS_ON {id: 12}]->(m2:Module {bar: 'five'})-[:DEPENDS_ON {id: 13}]->(m3:Module {bar: 'six'})");
+        
+        // with specific relationship
+        String customRelType = "DEPENDS_ON";
+        TestUtil.testResult(db, "MATCH (m1:Start) WITH collect(m1) as nodes CALL apoc.nodes.cycles(nodes, $config) YIELD path RETURN path",
+                map("config", map(REL_TYPE_KEY, customRelType)),
+                res -> {
+            List<Path> paths = Iterators.stream(res.<Path>columnAs("path"))
+                    .sorted(Comparator.comparingLong(item -> (long) item.lastRelationship().getProperty("id")))
+                    .collect(Collectors.toList());
+            assertEquals(2, paths.size());
+            assertionsCycle(paths.get(0), List.of("alpha", "one", "two", "alpha"), customRelType);
+            assertionsCycle(paths.get(1), List.of("alpha", "seven", "eight", "alpha"), customRelType);
+        });
+
+        // with all relationships
+        TestUtil.testResult(db, "MATCH (m1:Start) WITH collect(m1) as nodes CALL apoc.nodes.cycles(nodes) YIELD path RETURN path", res -> {
+            List<Path> paths = Iterators.stream(res.<Path>columnAs("path"))
+                    .sorted(Comparator.comparingLong(item -> (long) item.lastRelationship().getProperty("id")))
+                    .collect(Collectors.toList());
+            assertEquals(3, paths.size());
+            assertionsCycle(paths.get(0), List.of("alpha", "one", "two", "alpha"), null);
+            assertionsCycle(paths.get(1), List.of("alpha", "seven", "eight", "alpha"), null);
+            assertionsCycle(paths.get(2), List.of("beta", "three", "four", "beta"), null);
+        });
+
+        // with not existent relationship
+        TestUtil.testCallEmpty(db, "MATCH (m1:Start) WITH collect(m1) as nodes CALL apoc.nodes.cycles(nodes, $config) YIELD path RETURN path",
+                map("config", map(REL_TYPE_KEY, "NOT_EXISTENT")));
+        
+        // with maxDepth config
+        TestUtil.testCallEmpty(db, "MATCH (m1:Start) WITH collect(m1) as nodes CALL apoc.nodes.cycles(nodes, $config) YIELD path RETURN path", 
```

Comment:
Should we test something with self-rels? How does that work with this cycle procedure?

Reply:
Yep, added. It works also with self-rel (even if `maxDepth` config is 0)